hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndexWithSql.scala

```diff
@@ -103,4 +117,154 @@ class TestPartitionStatsIndexWithSql extends HoodieSparkSqlTestBase {
+           |  preCombineField = 'price',
+           |  hoodie.metadata.index.partition.stats.enable = 'true',
+           |  hoodie.metadata.index.column.stats.column.list = 'price'
+           |)
+           |location '$tablePath'
+           |""".stripMargin
+      )
+
+      spark.sql(
+        s"""
+           | insert into $tableName
+           | values (1, 'a1', 1000, 10), (2, 'a2', 2000, 20), (3, 'a3', 3000, 30), (4, 'a4', 2000, 10), (5, 'a5', 3000, 20), (6, 'a6', 4000, 30)
+           | """.stripMargin
+      )
+
+      // Validate partition_stats index exists
+      val metaClient = HoodieTableMetaClient.builder()
+        .setBasePath(tablePath)
+        .setConf(HoodieTestUtils.getDefaultStorageConf)
+        .build()
+      assertResult(tableName)(metaClient.getTableConfig.getTableName)
+      assertTrue(metaClient.getTableConfig.getMetadataPartitions.contains(PARTITION_STATS.getPartitionPath))
+
+      spark.sql("set hoodie.metadata.enable=true")
+      spark.sql("set hoodie.enable.data.skipping=true")
+      checkAnswer(s"select id, name, price, ts from $tableName where price>3000")(
+        Seq(6, "a6", 4000, 30)
+      )
+
+      // Test price update, assert latest value and ensure file pruning
+      spark.sql(s"update $tableName set price = price + 1 where id = 6")
+      checkAnswer(s"select id, name, price, ts from $tableName where price>3000")(
+        Seq(6, "a6", 4001, 30)
+      )
+
+      verifyFilePruning(
+        Map.apply(DataSourceReadOptions.ENABLE_DATA_SKIPPING.key -> "true", HoodieMetadataConfig.ENABLE.key -> "true"),
+        GreaterThan(AttributeReference("price", IntegerType)(), Literal(3000)),
+        HoodieTableMetaClient.reload(metaClient))
+    }
+  }
+
+  private def verifyFilePruning(opts: Map[String, String], dataFilter: Expression, metaClient: HoodieTableMetaClient): Unit = {
+    // with data skipping
+    val commonOpts = opts + ("path" -> metaClient.getBasePath.toString)
+    var fileIndex = HoodieFileIndex(spark, metaClient, None, commonOpts, includeLogFiles = true)
+    val filteredPartitionDirectories = fileIndex.listFiles(Seq(), Seq(dataFilter))
+    val filteredFilesCount = filteredPartitionDirectories.flatMap(s => s.files).size
+    assertTrue(filteredFilesCount < getLatestDataFilesCount(metaClient = metaClient))
+
+    // with no data skipping
+    fileIndex = HoodieFileIndex(spark, metaClient, None, commonOpts + (DataSourceReadOptions.ENABLE_DATA_SKIPPING.key -> "false"), includeLogFiles = true)
+    val filesCountWithNoSkipping = fileIndex.listFiles(Seq(), Seq(dataFilter)).flatMap(s => s.files).size
+    assertTrue(filesCountWithNoSkipping == getLatestDataFilesCount(metaClient = metaClient))
+  }
+
+  private def getLatestDataFilesCount(includeLogFiles: Boolean = true, metaClient: HoodieTableMetaClient) = {
+    var totalLatestDataFiles = 0L
+    getTableFileSystemView(metaClient).getAllLatestFileSlicesBeforeOrOn(metaClient.getActiveTimeline.lastInstant().get().getTimestamp)
```

Comment:
can we close the FSV :)

Reply:
good point.. closing both fsv and file index after usage.